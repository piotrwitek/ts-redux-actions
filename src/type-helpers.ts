/**
 * PUBLIC API
 */

/**
 * @desc Infers Action union-type from action-creator map object
 */
export type ActionType<
  ActionCreatorOrMap
> = ActionCreatorOrMap extends ActionCreator<StringType>
  ? ReturnType<ActionCreatorOrMap>
  : ActionCreatorOrMap extends object
  ? ActionCreatorMap<ActionCreatorOrMap>[keyof ActionCreatorOrMap]
  : never;

/**
 * @desc Infers State object from reducer map object
 */
export type StateType<ReducerOrMap> = ReducerOrMap extends (
  ...args: any[]
) => any
  ? ReturnType<ReducerOrMap>
  : ReducerOrMap extends object
  ? { [K in keyof ReducerOrMap]: StateType<ReducerOrMap[K]> }
  : never;

/**
 * INTERNAL API
 */

/**
 * @private
 * @desc Representing action-type of string
 */
export type StringType = string;

/**
 * @private
 * @desc Action without Payload
 * @type T - ActionType
 */
export type EmptyAction<T extends StringType> = {
  type: T;
};

/**
 * @private
 * @desc Action with only Payload
 * @type T - ActionType
 * @type P - Payload
 */
export type PayloadAction<T extends StringType, P> = {
  type: T;
  payload: P;
};

/**
 * @private
 * @desc Action with only Meta
 * @type T - ActionType
 * @type P - Payload
 * @type M - Meta
 */
export type MetaAction<T extends StringType, M> = {
  type: T;
  meta: M;
};

/**
 * @private
 * @desc Action with both Payload and Meta
 * @type T - ActionType
 * @type P - Payload
 * @type M - Meta
 */
export type PayloadMetaAction<T extends StringType, P, M> = {
  type: T;
  payload: P;
  meta: M;
};

/**
 * TODO: NOT USED
 * @private
 * @desc Flux Standard Action
 * @type T - ActionType
 * @type P - Payload
 * @type M - Meta
 */
export interface FluxStandardAction<T extends StringType, P = void, M = void> {
  type: T;
  payload: P;
  meta: M;
  error?: true;
}

/** @private */
export interface TypeMeta<T extends StringType> {
  getType?: () => T;
}

/** @private */
export type Box<T> = { v: T };

/** @private */
export type Unbox<T extends Box<any>> = T['v'];

/** @private */
export type NoArgCreator<T extends StringType> = () => EmptyAction<T>;

/** @private */
export type PayloadCreator<T extends StringType, P> = (
  payload: P
) => PayloadAction<T, P>;

/** @private */
export type PayloadMetaCreator<T extends StringType, P, M> = (
  payload: P,
  meta: M
) => PayloadMetaAction<T, P, M>;

/** @private */
export type FsaBuilder<
  T extends StringType,
  P extends Box<any> = Box<void>,
  M extends Box<any> = Box<void>
> = M extends Box<void>
  ? P extends Box<void>
    ? NoArgCreator<T>
    : PayloadCreator<T, Unbox<P>>
  : PayloadMetaCreator<T, Unbox<P>, Unbox<M>>;

/** @private */
export type FsaMapBuilder<
  T extends StringType,
  Result extends Box<any>,
  PArg extends Box<any> = Box<void>,
  MArg extends Box<any> = Box<void>
> = MArg extends Box<void>
  ? PArg extends Box<void>
    ? () => { type: T } & Unbox<Result>
    : (payload: Unbox<PArg>) => { type: T } & Unbox<Result>
  : (payload: Unbox<PArg>, meta: Unbox<MArg>) => { type: T } & Unbox<Result>;

/** @private */
export type ActionCreator<T extends StringType> = (
  ...args: any[]
) => { type: T };

/** @private */
export type ActionCreatorMap<T> = { [K in keyof T]: ActionType<T[K]> };
