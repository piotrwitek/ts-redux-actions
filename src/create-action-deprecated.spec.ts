import * as T from './type-helpers';
import { createActionDeprecated } from './create-action-deprecated';

describe('with symbol', () => {
  const CREATE_ACTION_DEPRECATED = Symbol(1);
  const increment = createActionDeprecated(CREATE_ACTION_DEPRECATED as any);
  // @dts-jest:pass:snap
  increment(); // => { type: CREATE_ACTION_DEPRECATED }
});

describe('with type only', () => {
  const withTypeOnly = createActionDeprecated('CREATE_ACTION_DEPRECATED');
  // @dts-jest:pass:snap
  withTypeOnly(); // => { type: 'CREATE_ACTION_DEPRECATED' }
});

describe('with payload', () => {
  const withPayload = createActionDeprecated(
    'CREATE_ACTION_DEPRECATED',
    (amount: number) => ({
      type: 'CREATE_ACTION_DEPRECATED',
      payload: amount,
    })
  );
  // @dts-jest:pass:snap
  withPayload(10); // => { type: 'CREATE_ACTION_DEPRECATED', payload: 10 }
});

describe('with meta', () => {
  const withMeta = createActionDeprecated(
    'CREATE_ACTION_DEPRECATED',
    (message: string) => ({
      type: 'CREATE_ACTION_DEPRECATED',
      meta: message,
    })
  );
  // @dts-jest:pass:snap
  withMeta('Error message'); // => { type: 'CREATE_ACTION_DEPRECATED', meta: 'Error message' }
});

describe('with payload and meta', () => {
  const withPayloadAndMeta = createActionDeprecated(
    'CREATE_ACTION_DEPRECATED',
    (username: string, message: string) => ({
      type: 'CREATE_ACTION_DEPRECATED',
      payload: { message: `${username}: ${message}` },
      meta: { username, message },
    })
  );
  // @dts-jest:pass:snap
  withPayloadAndMeta('Piotr', 'Hello!'); // => { type: 'CREATE_ACTION_DEPRECATED', payload: { message: 'Piotr: Hello!' }, meta: { username: 'Piotr', message: 'Hello!' } }
});
