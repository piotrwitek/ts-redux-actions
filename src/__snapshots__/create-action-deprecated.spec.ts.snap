// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`increment() 1`] = `"{ type: any; }"`;

exports[`withMeta('Error message') 1`] = `"{ type: \\"CREATE_ACTION_DEPRECATED\\"; meta: string; }"`;

exports[`withPayload(10) 1`] = `"{ type: \\"CREATE_ACTION_DEPRECATED\\"; payload: number; }"`;

exports[`withPayloadAndMeta('Piotr', 'Hello!') 1`] = `"{ type: \\"CREATE_ACTION_DEPRECATED\\"; payload: { message: string; }; meta: { username: string; message: string; }; }"`;

exports[`withTypeOnly() 1`] = `"{ type: \\"CREATE_ACTION_DEPRECATED\\"; }"`;
